services/mqtt/service.go

```diff
@@ -233,11 +242,81 @@ type handler struct {
 
 func (h *handler) Handle(event alert.Event) {
 	h.diag.HandlingEvent()
-	if err := h.s.Alert(h.c.BrokerName, h.c.Topic, h.c.QoS, h.c.Retained, event.State.Message); err != nil {
+	topic, err := h.renderTopic(h.c.Topic, event)
+	if err != nil {
+		h.diag.Error("failed to create MQTT topic from template", err)
+	} else if err = h.s.Alert(h.c.BrokerName, topic, h.c.QoS, h.c.Retained, event.State.Message); err != nil {
 		h.diag.Error("failed to post message to MQTT broker", err)
 	}
 }
 
+type idInfo struct {
+	// Measurement name
+	Name string
+
+	// Task name
+	TaskName string
+
+	// Concatenation of all group-by tags of the form [key=value,]+.
+	// If not groupBy is performed equal to literal 'nil'
+	Group string
+
+	// Map of tags
+	Tags map[string]string
+}
+
+type messageInfo struct {
+	idInfo
+
+	// The ID of the alert.
+	ID string
+
+	// Fields of alerting data point.
+	Fields map[string]interface{}
+
+	// Alert Level, one of: INFO, WARNING, CRITICAL.
+	Level string
+
+	// Time
+	Time time.Time
+
+	// Duration of the alert
+	Duration time.Duration
+}
```

Comment:
These structures should not be copied. Use the `TemplateData` function on the alert Event in order to use consistent template data for the templates.

See https://github.com/influxdata/kapacitor/blob/master/services/pagerduty2/service.go#L402 as an example.

Reply:
Done.